File: CommandPython/StreetLightCommand.py
# Copyright (c) 2022 RodKingroo

from StreetLight import *
from Command import *

def setUndo(streetLight, prevBright):
    if prevBright == streetLight.STAGE_ONE:
        streetLight.stage_1()
    elif prevBright == streetLight.STAGE_TWO:
        streetLight.stage_2()
    elif prevBright == streetLight.STAGE_THREE:
        streetLight.stage_3()
    elif prevBright == streetLight.STAGE_FOUR:
        streetLight.stage_4()
    elif prevBright == streetLight.STAGE_FIVE:
        streetLight.stage_5()
        
# Light on Stage One command
class LightOnStageOneCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.on()
        self.streetLight.stage_1()
        
    def undo(self):
        self.streetLight.off()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        

# Light on Stage Two command
class LightOnStageTwoCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.on()
        self.streetLight.stage_2()
        
    def undo(self):
        self.streetLight.off()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        
        
# Light on Stage Three command
class LightOnStageThreeCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.on()
        self.streetLight.stage_3()
        
    def undo(self):
        self.streetLight.off()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        
        
        
# Light on Stage Four command
class LightOnStageFourCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.on()
        self.streetLight.stage_4()
        
    def undo(self):
        self.streetLight.off()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        
        
# Light on Stage Five command
class LightOnStageFiveCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.on()
        self.streetLight.stage_5()
        
    def undo(self):
        self.streetLight.off()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        
# Light Off command
class LightOffCommand(Command):
    def __init__(self, streetLight):
        if isinstance(streetLight, StreetLight):
            self.streetLight = streetLight
        else:
            raise NotImplementedError
        
    def execute(self):
        self.prevBright = self.streetLight.getState()
        self.streetLight.off()
        
    def undo(self):
        self.streetLight.on()
        self.streetLight = setUndo(self.streetLight, self.prevBright)
        

File: CommandCPP/StreetLightCommand.cpp
/* Copyright (c) 2022 RodKingroo */
#include "StreetLightCommand.h"

void setUndo(shared_ptr<StreetLight> light, int prevBright)
{
    switch (prevBright)
    {
        case StreetLight::OFF:
            light->off();
            break;
        case StreetLight::STATE_ONE:
            light->state_1();
            break;
        case StreetLight::STATE_TWO:
            light->state_2();
            break;
        case StreetLight::STATE_THREE:
            light->state_3();
            break;
        case StreetLight::STATE_FOUR:
            light->state_4();
            break;
        case StreetLight::STATE_FIVE:
            light->state_5();
            break;

    }
}


/* Street Light On State One Command */
StreetLightOnStateOneCommand::StreetLightOnStateOneCommand(shared_ptr<StreetLight> light)
{
    this->light = light;
}

void StreetLightOnStateOneCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_1();
}

void StreetLightOnStateOneCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}



/* Street Light on State Two Command */
StreetLightOnStateTwoCommand::StreetLightOnStateTwoCommand(shared_ptr<StreetLight> light)
{
    this->light = light;

}

void StreetLightOnStateTwoCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_2();

}

void StreetLightOnStateTwoCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Street Light on State Three Command */
StreetLightOnStateThreeCommand::StreetLightOnStateThreeCommand(shared_ptr<StreetLight> light)
{
    this->light = light;
}

void StreetLightOnStateThreeCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_3();

}

void StreetLightOnStateThreeCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Street Light on State Three Command */
StreetLightOnStateFourCommand::StreetLightOnStateFourCommand(shared_ptr<StreetLight> light)
{
    this->light = light;
}

void StreetLightOnStateFourCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_4();
}

void StreetLightOnStateFourCommand::undo() /* override */
{
    light->off();  
    setUndo(light, prevBright);
}


/* Street Light on State Five Command */
StreetLightOnStateFiveCommand::StreetLightOnStateFiveCommand(shared_ptr<StreetLight> light)
{
    this->light = light;
}

void StreetLightOnStateFiveCommand::execute() /* override */
{
    prevBright = light->getState();
    light->on();
    light->state_5();
}

void StreetLightOnStateFiveCommand::undo() /* override */
{
    light->off();
    setUndo(light, prevBright);
}


/* Street Light on Off Command */
StreetLightOffCommand::StreetLightOffCommand(shared_ptr<StreetLight> light)
{
    this->light = light;
}

void StreetLightOffCommand::execute() /* override */
{
    prevBright = light->getState();
    light->off();
}

void StreetLightOffCommand::undo() /* override */
{
    light->on();
    setUndo(light, prevBright);
}